### ISSUE TITLE
Support for human data collection

### ISSUE THREAD

**Issue by mpomarlan** [ease-crc/soma#140]
I have already created a branch (human-neems) to take care of the issues mentioned below. This branch is going to be somewhat "experimental" and loose, because we want to have something to show for the milestone, but the idea is to clean and merge it into master at some point.

The priority issues to tackle:

1) represent "meta-data" about an episode, i.e. what is being done. Example: "set the table for a formal lunch for four people".

2) embed or link raw data, such as videos, to a NEEM.

3) concepts and labels for objects

**Comment by mpomarlan** [comment]
Issue 3 is "taken care of" by the latest commit to the human-neems branch.

I defined several new object concepts, and ELAN labels for them.

An interesting feature: some object concepts appear only in SOMA-ELAN as opposed to SOMA-HOME. These are concepts for objects which are a) containers and b) intended to contain a particular substance. An example is WineBottle.

The axiomatization isn't there at the moment, but the idea for something like WineBottle is that it is equivalent to Bottle and contains some Wine (or, alternatively, "is intended" to contain wine).

The reason for this classification: people use it. In particular, our human-data collection researchers use it. It makes sense for them, because they record data for one episode, during which the use of a WineBottle won't change. So what they are saying with colloquial language, in which we humans define a container by its contentes, is that for the duration of the episode, a particular container held the same kind of content.

Note that such mixed concepts like WineBottle will appear only in SOMA-ELAN.

**Comment by daniel86** [comment]
> I have already created a branch (human-neems) to take care of the issues mentioned below. This branch is going to be somewhat "experimental" and loose, because we want to have something to show for the milestone, but the idea is to clean and merge it into master at some point.
> 
> The priority issues to tackle:
> 
>     1. represent "meta-data" about an episode, i.e. what is being done. Example: "set the table for a formal lunch for four people".
> 
>     2. embed or link raw data, such as videos, to a NEEM.
> 
>     3. concepts and labels for objects

2. raw data is stored as time series data, the association to events is done via the time. No explicit link is there between data and entities. For now, only the NEEM meta data may have fields that assign additional data sources to entities -- e.g. that additional information about an object can be found in a URDF file.

**Comment by daniel86** [comment]
> Issue 3 is "taken care of" by the latest commit to the human-neems branch.
> 
> I defined several new object concepts, and ELAN labels for them.
> 
> An interesting feature: some object concepts appear only in SOMA-ELAN as opposed to SOMA-HOME. These are concepts for objects which are a) containers and b) intended to contain a particular substance. An example is WineBottle.
> 
> The axiomatization isn't there at the moment, but the idea for something like WineBottle is that it is equivalent to Bottle and contains some Wine (or, alternatively, "is intended" to contain wine).
> 
> The reason for this classification: people use it. In particular, our human-data collection researchers use it. It makes sense for them, because they record data for one episode, during which the use of a WineBottle won't change. So what they are saying with colloquial language, in which we humans define a container by its contentes, is that for the duration of the episode, a particular container held the same kind of content.
> 
> Note that such mixed concepts like WineBottle will appear only in SOMA-ELAN.

1. We already have a first meta information schema for neems, basically the meta information is stored in an additional Mongo DB collection, but not as triples.

**Comment by mpomarlan** [comment]
On point 2: so if I would like to attach a video, how would that be done? The link to the video file wouldn't itself be a time series data. Would this be done via the meta info, i.e. link some event/episode to the video file?

**Comment by daniel86** [comment]
> On point 2: so if I would like to attach a video, how would that be done? The link to the video file wouldn't itself be a time series data. Would this be done via the meta info, i.e. link some event/episode to the video file?

good point. We haven't really fixed how to handle video data yet.
I once actually stored video frames in mongo DB, and could reconstruct the video from it :)
But might not be optimal. And best way to store video might very well be a file format with some compression of the data.
Still, video is also kind-of time data and you could show a frame that e.g. corresponds to the start of an event.

If it is just about providing an additional video for a neem, I think, the video should just be uploaded as part of the NEEM which makes it accessible in openEASE. But we would not need to model it.

**Comment by mpomarlan** [comment]
Ok, so at the moment the plan is, as far as I understand, to use a metadata.json file to address points 1 (a brief description of the episode) and 2 (linking raw data to entities), and there is substantial freedom for me and Moritz as to how to do this.

All right, but for the milestone presumably it will be nice to show how the NEEM can also be queried. These queries might involve the metadata, which Moritz and I are about to invent. How can we write the predicates on the NEEMhub side to take care of these queries?

**Comment by daniel86** [comment]
> Ok, so at the moment the plan is, as far as I understand, to use a metadata.json file to address points 1 (a brief description of the episode) and 2 (linking raw data to entities), and there is substantial freedom for me and Moritz as to how to do this.
> 
> All right, but for the milestone presumably it will be nice to show how the NEEM can also be queried. These queries might involve the metadata, which Moritz and I are about to invent. How can we write the predicates on the NEEMhub side to take care of these queries?

could you provide an example of what you intent to do?

There is a NEEM filtering functionality in openEASE which allows to query all NEEMs that fulfill some criterion, I guess this is not what you want?

Otherwise the meta data is rather supposed to be used to display the NEEM in openEASE, which is still work in progress because we haven't fixed what the meta data of a NEEM is. But @seba90  should be involved in this, as he would also need to provide complete meta data in generated robot NEEMs

**Comment by mpomarlan** [comment]
I'll talk to Moritz asap about what queries would be nice to showcase for the milestone. At the latest, there is an H meeting tomorrow at noon.

**Comment by mpomarlan** [comment]
Ok, I think for milestone showcase purposes it boils down to two kinds of queries:

1) return episode (ids) for all episodes containing a particular action with a particular participant (e.g. putting down a glass)

2) return data of a particular modality associated to an episode (e.g. return a link to a video file associated with the episode)

Query 1) should work as-is, I think, because it only uses information in the triples. Query 1) should be answerable for robot neems, so if the human neems use the same sort of triples this should work without extra effort.

Query 2) will need implementation. What's going to happen is that Moritz will define some entries in the metadata.json file, among which are links to files with raw data of various modalities. Basically, given the episode and modality, the query should return one of those links.

**Comment by daniel86** [comment]
to be honest there is a difficulty with 1) as different neems are stored in different collections in mongo. So we could write a query like this in KnowRob, but it would be very slow as the query would need to be executed once for each collection. Could be that there is a way to make it efficient, but I am not aware of it.

**Comment by daniel86** [comment]
about 2) who should return the link? where should it be returned and to whom? I am still unsure about the usecase as you will see all the data files associated to the neem easily in openEASE, you can e.g. click on the video and watch it. no query needed

**Comment by mpomarlan** [comment]
One use case is someone from the H group assembling a training data set for some neural network. Ideally, this person might have a program that calls the NEEMhub with a query which selects the data they are interested in, feeds it back into their program, and has deep learning do its magic.

For the purpose of just showing the data is there, sure, as long as you identify an episode via something like query 1, then fine, a user can click. But to really make this ergonomic for big data applications, you must have queries able to assemble data from several NEEMs without the user having to click download all the time.

**Comment by mpomarlan** [comment]
About query 1, it is pretty much a must. Felix wants to present how episodes containing glasses being put down are identifiable in the NEEM hub, since all H subgroups have episodes matching this query and then one can show the variety of sources feeding into the NEEM hub.

**Comment by daniel86** [comment]
> About query 1, it is pretty much a must. Felix wants to present how episodes containing glasses being put down are identifiable in the NEEM hub, since all H subgroups have episodes matching this query and then one can show the variety of sources feeding into the NEEM hub.

yes we can do something like that. But I would tend to do it outside of the knowledge base. We have this episode filter functionality in openEASE. Maybe it would make sense to extend it?
Then the queried can be issued in openEASE, and the NEEM discovery page would only show the matching neems.
But that also would require some implementation work in openEASE to be done

**Comment by daniel86** [comment]
> One use case is someone from the H group assembling a training data set for some neural network. Ideally, this person might have a program that calls the NEEMhub with a query which selects the data they are interested in, feeds it back into their program, and has deep learning do its magic.
> 
> For the purpose of just showing the data is there, sure, as long as you identify an episode via something like query 1, then fine, a user can click. But to really make this ergonomic for big data applications, you must have queries able to assemble data from several NEEMs without the user having to click download all the time.

we have infrastructure for that. This is all handled on hadoop. Ask @seba90  about it.
This is not really our business, but the responsibility of Asil and Sebastian.
So Models are learned automatically when NEEMs are added to the HUB. How to set this up? I do not know, but seba or asil should know it.

**Comment by seba90** [comment]
The intention is that the NEEM Hub will do learn automatically it in the future, however this not the case now and will not be integrated until the milestone. As explained in the H talks, I am using KnowRob to convert those neems into CSV files which  are then used to train my models.

**Comment by daniel86** [comment]
Ok, fine. But I wouldn't want to create a temporary interface for this. Also because many other important things are still pending.
So I guess, Moritz would need to do it like seba for now, and write his own conversion script.
Or could Moritz maybe use your script @seba90  ?

**Comment by seba90** [comment]
He would need Knowrob to do that. https://github.com/code-iai/narrative2vec here can be my conversion script found.

**Comment by mpomarlan** [comment]
Ok so for me to have a summary of what could happen/be done:

NEEMs from human data collection can be uploaded to the NEEM hub
using openEASE it is possible to retrieve a list of NEEMs from episodes where a glass was put down; this predicate might need some work to implement though
once retrieved in openEASE, it is possible for a human user to access any data linked to the NEEM, e.g. to download and play a video

Is this correct?

**Comment by daniel86** [comment]
> Ok so for me to have a summary of what could happen/be done:
> 
> NEEMs from human data collection can be uploaded to the NEEM hub
> using openEASE it is possible to retrieve a list of NEEMs from episodes where a glass was put down; this predicate might need some work to implement though
> once retrieved in openEASE, it is possible for a human user to access any data linked to the NEEM, e.g. to download and play a video
> 
> Is this correct?

well the second part is also not there yet, as the NEEM-Hub does not provide us yet with an interface to retrieve data files. But @seba90  promised to provide some interface for that today

**Comment by mpomarlan** [comment]
Good. So a possible demo from the H-area might be, someone types a query about finding episodes where a glass is put down, the query retrieves NEEMs from each of the H area subprojects, and the user can play videos from each NEEM to show the different sources and methods of capture.

Some NEEMs may have time series data for position as well btw. I talked with Johannes Pfau and Janis Roßkamp yesterday to give them a brief intro into tf records. If they use the same kind of entries then presumably displaying this position data for their NEEMs won't be too painful.

**Comment by seba90** [comment]
There is no concept right now how to handle the "attached" files to NEEMs. I would assume those have to be stored in the mongo db and there is then an link in the triple store. Like images taken by the robot. Otherwise the H group needs to upload the videos in the raw section (I  sent an email to the h group regarding that) and openEASE can download it via dvc get (https://dvc.org/doc/start/data-access).

**Comment by mpomarlan** [comment]
With the Begehung behind us, this issue is now no longer relevant. Of course, supporting human NEEMs remains important, but this discussion was about some quick and temporary measures to support the review. Since the review is over, this discussion can conclude.
